pkg/runtime/depot.go

```diff
@@ -300,6 +331,46 @@ func (d *depot) validateVolume(absoluteDest string) error {
 	return nil
 }
 
+// getSharedFilesToReLink returns a map of deployed files to re-link to (or re-copy from) another
+// artifact that provides those files. The key is the deployed file path and the value is the
+// source path from another artifact.
+func (d *depot) getSharedFilesToReLink(id strfmt.UUID, relativeSrc, path string) (map[string]string, error) {
+	// Map of deployed paths to other sources that provides those paths.
+	relink := make(map[string]string, 0)
+
+	// Get a listing of all files deployed by this artifact.
+	deployedDir := filepath.Join(d.Path(id), relativeSrc)
+	deployedFiles, err := fileutils.ListDirSimple(deployedDir, false)
+	if err != nil {
+		return nil, errs.Wrap(err, "failed to list depot files for artifact")
+	}
+
+	// For each of those files, find another artifact (if any) that deploys its own copy.
+	for _, deployedFile := range deployedFiles {
+		relativeDeployedFile := deployedFile[len(deployedDir)+1:]
+		for artifactId, artifactDeployments := range d.config.Deployments {
+			if artifactId == id {
+				continue
+			}
+
+			for _, deployment := range artifactDeployments {
+				for _, relativeDeploymentFile := range deployment.Files {
+					if relativeDeployedFile == relativeDeploymentFile {
+						// We'll want to relink this other artifact's copy after undeploying the currently deployed version.
+						deployedFile := filepath.Join(path, relativeDeployedFile)
+						newSrc := filepath.Join(d.Path(artifactId), deployment.InstallDir, relativeDeployedFile)
+						logging.Debug("More than one artifact provides '%s'", relativeDeployedFile)
+						logging.Debug("Will relink %s to %s", deployedFile, newSrc)
+						relink[deployedFile] = newSrc
```

Comment:
Should probably break here. What if we have three artifacts linking to the same file? We still want to address the first one encountered.

Reply:
I don't see the point of this. Deployment iteration is not ordered, so we cannot make any sort of guarantee which of 2 remaining artifacts will get picked.

Furthermore, we'd have to break out of two for loops, not one, so that would either require a `goto` or refactor to use functions. I don't think it's worth it since we cannot make the guarantee mentioned earlier.

Reply:
Not being able to make a guarantee is already the reality we live in now. That is a problem for another day that we cannot address in this story. All we can address is that as long as there are artifacts still targeting a file, that file should exist. And one of them will arbitrarily win the race.

The main reason I brought this up is because as currently written we would remove and then link the file again here: 

https://github.com/ActiveState/cli/pull/3455/files#diff-e57fe7a86f0cf670f9bb15dec637e18669380299ce1d661b3872f4e822c65bf2R299

Reply:
Okay, I added a `goto` to break out of 3 `for` loops.